tools/ci-trigger/pr.go

```diff
@@ -0,0 +1,324 @@
+		if err := prototext.Unmarshal(in, md); err != nil {
+			return err
+		}
+		for _, d := range virtualDeviceTypes {
+			// We can't support virtual ATE tests - skip them.
+			if strings.Contains(ft, "ate_tests") {
+				continue
+			}
+			badgeTestName := base64.RawURLEncoding.EncodeToString([]byte(ft))
+			deviceName := strings.ReplaceAll(d.String(), " ", "_")
+			badgePath := gcpBucketPrefix + "/" + strconv.Itoa(p.ID) + "/" + p.HeadSHA + "/" + badgeTestName + "." + deviceName + ".svg"
+			badgeURL := "https://storage.googleapis.com/" + gcpBucket + "/" + badgePath
+			newTest := functionalTest{
+				Name:        md.PlanId,
+				Description: md.Description,
+				Path:        ft,
+				DocURL:      "https://github.com/" + githubProjectOwner + "/" + githubProjectRepo + "/blob/" + p.HeadSHA + "/" + ft + "/README.md",
+				TestURL:     "https://github.com/" + githubProjectOwner + "/" + githubProjectRepo + "/blob/" + p.HeadSHA + "/" + ft,
+				BadgeURL:    badgeURL,
+				BadgePath:   badgePath,
+				Status:      "pending authorization",
+			}
+			tests[d] = append(tests[d], newTest)
+		}
+	}
+
+	for _, d := range virtualDeviceTypes {
+		if dt, ok := tests[d]; ok {
+			p.Virtual = append(p.Virtual, device{
+				Type:  d,
+				Tests: dt,
+			})
+		}
+	}
+
+	return nil
+}
+
+// firstComment returns the first bot-originated comment in the PR if it exists.
+func (p *pullRequest) firstComment(ctx context.Context, githubClient *github.Client) (*github.IssueComment, error) {
+	comments, _, err := githubClient.Issues.ListComments(ctx, githubProjectOwner, githubProjectRepo, p.ID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, comment := range comments {
+		if comment.GetUser().GetLogin() == githubBotName {
+			return comment, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// functionalTestPaths returns a list of directories containing functional test metadata.
+func functionalTestPaths(f fs.FS) ([]string, error) {
+	var testdirs []string
+	err := fs.WalkDir(f, "feature", func(path string, d fs.DirEntry, err error) error {
+		if !d.IsDir() && d.Name() == "metadata.textproto" {
+			testdirs = append(testdirs, filepath.Dir(path))
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return testdirs, nil
+}
+
+// modifiedFunctionalTests checks if any values of modifiedFiles starts with
+// functionalTests.  The list returned contains functional test paths with at
+// least one modified file.
+func modifiedFunctionalTests(functionalTests []string, modifiedFiles []string) []string {
+	fts := make(map[string]struct{})
+	for _, ft := range functionalTests {
+		for _, mf := range modifiedFiles {
+			if strings.HasPrefix(mf, ft) {
+				fts[ft] = struct{}{}
+			}
+		}
+	}
+	result := []string{}
+	for k := range fts {
+		result = append(result, k)
+	}
+	return result
+}
+
+// newPullRequest returns a minimally-populated pull request.
+func newPullRequest(id int, cloneURL string, headSHA string, baseSHA string, tmpDir string) *pullRequest {
```

Comment:
Not sure this pulls it's weight. I'd let callers create it directly, especially since direct struct creation avoids potential bugs in transposing all the string arguments.